### ISSUE TITLE
[doc] CMake flag for building Doxygen reference guide

### ISSUE THREAD

**Issue by ferdymercury** [root-project/root#8950]
### Explain what you would like to see improved
As a follow-up of https://github.com/root-project/root/issues/8947, it would be nice to have a CMake flag that allows you to turn on the building of the Doxygen documentation within a normal build. Right now, it has to be done manually in the source directory.

### Optional: share how it could be improved
In addition, cmake Options can be easily used to 'tune' the input files, see https://aliceo2group.github.io/advanced/doxygen.html, rather than manually having to edit the Doxyfile as indicated in https://root.cern/for_developers/doxygen/#how-to-generate-the-doxygen-documentation.

Some flags like:
```
BUILD_DOCS [OFF]
BUILD_DOCS_CORE [ON]
BUILD_DOCS_GRAF2D [OFF]
...
```
and then calling:

`doxygen_add_docs(...)` with all the variables correctly set to generate the proper Doxyfile. This requires some manual copy work from the actual Doxyfile to the CMakeLists.txt, but it will look clean in the end.

Alternatively, one could just modify-on-the-fly the preexisting Doxyfile file as follows:

```
    set(doxyfile_in ${CMAKE_CURRENT_SOURCE_DIR}/Doxyfile)
    set(doxyfile ${CMAKE_CURRENT_BINARY_DIR}/Doxyfile)
    configure_file(${doxyfile_in} ${doxyfile} @ONLY)
    add_custom_command(
      OUTPUT ${CMAKE_CURRENT_BINARY_DIR}/doxygen.stamp
      DEPENDS ${doxyfile} ${HEADER_FILES}
      WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
      COMMAND sed "s;INPUT                  =;INPUT                  = ${MY_NEW_INPUT_FILES} ${doxyfile} | ${DOXYGEN_EXECUTABLE} -
      COMMAND cmake -E touch ${CMAKE_CURRENT_BINARY_DIR}/doxygen.stamp
      COMMENT "Generating API documentation with Doxygen"
      VERBATIM)
```

### To Reproduce
https://root.cern/for_developers/doxygen/#how-to-generate-the-doxygen-documentation

### Setup
```
   ------------------------------------------------------------------
  | Welcome to ROOT 6.25/01                        https://root.cern |
  | (c) 1995-2021, The ROOT Team; conception: R. Brun, F. Rademakers |
  | Built for linuxx8664gcc on May 17 2021, 10:49:42                 |
  | From heads/master@v6-25-01-861-g5ea261143f                       |
  | With                                                             |
  | Try '.help', '.demo', '.license', '.credits', '.quit'/'.q'       |
   ------------------------------------------------------------------
```

### Additional context
https://github.com/root-project/root/issues/8781

**Comment by couet** [comment]
@ferdymercury : that's a good point. That will have to be discussed with the team and in particular wit @bellenot who is responsible of came in root. It is quite a big change seems to me.

**Comment by couet** [comment]
Recent changes in the way the list of directories to be build might obsolete this issue. The documentation build is not part of the cmake procedure. It has its own Makefile. Now the list of directories is build via  script. The page  
https://root.cern/for_developers/doxygen/#how-to-generate-the-doxygen-documentation
has been. updated.

**Comment by ferdymercury** [comment]
Thanks for the feedback. Just as a suggestion, I think it would still be possible to migrate from that Makefile to a CMakeLists.txt

That way, one could configure via the command line what part of the docs to build, instead of having to modify the makeinput.sh script, which in turn modifies the git source directory, which brings us back to https://github.com/root-project/root/issues/8947

**Comment by couet** [comment]
I am now looking at #8947. I am modifying the `Makefile` to left a clean source directory after build. `makeinput.sh` should be modified only temporarily by developers to make only part of the doc in order to speed the build when working on a specific part of the reference guide.  This script is not only a static list. At the end is added the pieces of doc build by `xtract_docstrings.py`
 and `print_roofit_pyz_doctrings.py`. But it is true that if one modifies (even temporarily)  `makeinput.sh` then it will appears in `git status`. I am not should how a `cmake`  approach will fix that.

**Comment by ferdymercury** [comment]
Thanks. What I am suggesting is something like:
```
./configure --enable-th1 --enable-th2
make
```
which lets you tune without modifying the source dir. Or the alternative with cmake, which would be:
```
cmake -DENABLE_TH1 -DENABLE_TH2
make
```
With cmake it would be a bit cleaner, as there is the function doxygen_add_docs() which already does what makeinput.sh is doing manually. See Alice: https://github.com/AliceO2Group/Monitoring/blob/dev/doc/CMakeLists.txt#L22
But with ./configure it should probably be doable too.

**Comment by couet** [comment]
Yes but the list of folders does not follow the classes naming .... You may need ore `-DENABLE_HIST` .... but then is the `hist` folder we have `hist` and `histpainter` subfolders which might be enabled separately ... and is it a simple example there is much more complex structure for with we need a precise picking which is done by  `makeinput.sh`

**Comment by ferdymercury** [comment]
True... But the cherry picking might be still doable by hacking the CMakeLists.txt (instead of hacking the makeinput.sh as right now). But yeah, we don't win much in that case.

**Comment by couet** [comment]
Yes because `CMakeLists.txt` will appear as modified instead of `makeinput.sh`. Anyway there is some clean up to be done in order to avoid the sources polluting. I am looking at it..
